finality-aleph/src/data_io.rs

```diff
@@ -178,15 +176,13 @@ where
         let mut import_stream = self.client.import_notification_stream();
         loop {
             tokio::select! {
-                Some(message) = &mut self.network.next() => {
+                Some(message) = async {
+                    let mut lock = self.messages_from_network.lock().await;
+                    lock.next().await
```

Comment:
The component is `Arc`, so it can be shared in multiple places, which means that there may be another place in the code that calls `lock.next()` - that's a race condition for the message.

Reply:
Isn't it that indeed many threads can run line 180 at the same time, but only one will be granted lock so line 181 is in critical section?

Reply:
I meant that you pass in the `ComponentNetwork` instance to the constructor, and create an instance of `recipient` (`messages_from_network`). If that same instance of `ComponentNetwork` is passed somewhere else, that place also can call `component_network.receiver()` and create an instance of a `receiver` which will compete for messages with yours.

Reply:
But `component_network` is passed with ownership  to `DataStore::new`, hence there cannot be other place in the code that uses `component_network` reference after that. It must have been `component_network.clone()`, assuming `ComponentNetwork` can be `clone()`d, and it that case we should be good?

Reply:
Actually you can have multiple `Arc`es of the receiver, since you could've gotten some before handing away the `ComponentNetwork`. But they receiver itself is behind a `Mutex` within the `Arc`, so cloning the `Arc` does not matter. In fact this is the usual way in which you use mutexes in Rust.

Reply:
> But they receiver itself is behind a `Mutex` within the `Arc`, so cloning the `Arc` does not matter.

But that `Mutex` will be dropped at some point and the other thread could acquire an access to it.

Reply:
Ohh, I see, you mean some other thread might get a specific message instead of this one? Yeah, that is correct, depending on what we are doing this might even be intended (see `network/split.rs` for an example). In this case we don't want that to happen, but we have to ensure that ourselves.

Reply:
Yes, that's exactly what I meant :)